### PR TITLE
bpo-1635741: Port symtable module to multiphase initialization

### PR DESCRIPTION
**Pull request by tiran**
Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue1635741

Automerge-Triggered-By: GH:tiran

### CODE REVIEW

File: Modules/symtablemodule.c [hunk 1]
```diff
@@ -71,13 +71,57 @@ static PyMethodDef symtable_methods[] = {
     {NULL,              NULL}           /* sentinel */
 };
 
+static int
+symtable_init_stentry_type(PyObject *m)
+{
+    return PyType_Ready(&PySTEntry_Type);
```
Comment (vstinner) [inline Modules/symtablemodule.c:77]: I don't see the value of having two exec slots. Why not simply putting all exec code into a single function?

File: Modules/symtablemodule.c [hunk 2]
```diff
@@ -71,13 +71,57 @@ static PyMethodDef symtable_methods[] = {
...
+
 static struct PyModuleDef symtablemodule = {
     PyModuleDef_HEAD_INIT,
     "_symtable",
     NULL,
-    -1,
+    0,
     symtable_methods,
-    NULL,
+    symtable_slots,
```
Comment (vstinner) [inline Modules/symtablemodule.c]: Would you mind to use C99 designated initializers? ._name = ..., .m_methods = ..., etc.

File: Modules/symtablemodule.c [hunk 3]
```diff
@@ -71,13 +71,57 @@ static PyMethodDef symtable_methods[] = {
...
 
+static int
+symtable_init_stentry_type(PyObject *m)
+{
+    return PyType_Ready(&PySTEntry_Type);
+}
+
+static int
+symtable_init_constants(PyObject *m)
+{
```
Comment (vstinner) [inline Modules/symtablemodule.c:82]: I'm trying to slowly use macros to factorize the code, something like:

```
#define ADD_INT(value) \
    if (PyModule_AddIntConstant(module, #value, value) < 0) { \
        return -1; \
    }
```

A more generic machinery like the one you proposed would avoid such macro, and until there is such API, IMO using a macro makes the code more readable:

```
ADD_INT(USE);
```

You can add a second macro for PyModule_AddIntConstant() calls.

File: Modules/symtablemodule.c (hunk 1 above)
Comment (tiran) [reply Modules/symtablemodule.c:77]: They do different things. Also ``PySTEntry_Type`` is define somewhere else and we might want to move the type initialization when the ``PySTEntry_Type`` is refactored.

File: Modules/symtablemodule.c (hunk 3 above)
Comment (tiran) [reply Modules/symtablemodule.c:82]: I'd rather not spend more time then necessary. This was a simple search and replace job.

Let's consider refactoring after ``PyModuleConstants_Def`` from https://github.com/python/cpython/pull/23286 has landed. The new feature will make module constant definitions easier.

File: Modules/symtablemodule.c (hunk 2 above)
Comment (tiran) [reply Modules/symtablemodule.c]: Fixed

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**